### PR TITLE
Add admin sdk

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ project(realm)
 
 # Project-wide build flags
 set(CMAKE_CXX_STANDARD 20)
+#set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address")
```
Comment (leemaguire) [inline CMakeLists.txt:7]: Why is this commented out?

File: CMakeLists.txt [hunk 2]
```diff
@@ -105,14 +106,31 @@ if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL
...
+find_library(FoundationLib CoreFoundation)
+message("LIB: ${FoundationLib}")
+if (APPLE)
+    list(APPEND CMAKE_SYSTEM_FRAMEWORK_PATH
+            ${CMAKE_OSX_SYSROOT_PATH}/Library/Frameworks
+            ${CMAKE_OSX_SYSROOT_PATH}/Network/Library/Frameworks
+            ${CMAKE_OSX_SYSROOT_PATH}/System/Library/Frameworks
+            )
+    target_link_libraries(cpprealm "-framework CoreFoundation")
+    #    target_link_libraries(cpprealm_exe_tests "-framework CoreFoundation")
```
Comment (leemaguire) [inline CMakeLists.txt:125]: This line can be removed

File: CMakeLists.txt [hunk 3]
```diff
@@ -105,14 +106,31 @@ if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL
...
 
-target_link_libraries(cpprealm Storage ObjectStore Sync z curl)
+target_link_libraries(cpprealm Storage ObjectStore Sync z)
 
 set_target_properties(cpprealm PROPERTIES PUBLIC_HEADER "${HEADERS}")
 
 install(TARGETS cpprealm
         PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/cpprealm)
+set(CMAKE_OSX_SYSROOT ${CMAKE_OSX_SYSROOT_PATH})
+find_library(FoundationLib CoreFoundation)
```
Comment (leemaguire) [inline CMakeLists.txt:116]: this should be inside `if(APPLE)`

File: src/cpprealm/app.hpp [hunk 4]
```diff
@@ -164,14 +171,102 @@ app::Response do_http_request(const app::Request& request)
...
+        }
+        CFReadStreamRef myReadStream = CFReadStreamCreateForHTTPRequest(kCFAllocatorDefault, myRequest);
+        CFReadStreamOpen(myReadStream);
+        auto error = CFReadStreamGetError(myReadStream);
+        if (error.error) {
+            abort();
+        }
+        std::string response_string;
+
+//        CFIndex bytes_read;
```
Comment (leemaguire) [inline src/cpprealm/app.hpp:240]: remove commented out code

File: src/cpprealm/app.hpp [hunk 5]
```diff
@@ -164,14 +171,102 @@ app::Response do_http_request(const app::Request& request)
...
 public:
     void send_request_to_server(app::Request&& request,
                                 util::UniqueFunction<void(const app::Response&)>&& completion_block)
     {
+
         completion_block(do_http_request(request));
     }
 };
+#else
+class DefaultTransport : public app::GenericNetworkTransport {
```
Comment (leemaguire) [inline src/cpprealm/app.hpp:186]: Should all be snake_cased

File: src/cpprealm/app.hpp [hunk 6]
```diff
@@ -164,14 +171,102 @@ app::Response do_http_request(const app::Request& request)
...
+                CFHTTPMessageSetHeaderFieldValue(myRequest, name, value);
+                CFRelease(name);
+                CFRelease(value);
+            }
+        }
+        CFReadStreamRef myReadStream = CFReadStreamCreateForHTTPRequest(kCFAllocatorDefault, myRequest);
+        CFReadStreamOpen(myReadStream);
+        auto error = CFReadStreamGetError(myReadStream);
+        if (error.error) {
+            abort();
```
Comment (leemaguire) [inline src/cpprealm/app.hpp:236]: Might be better to throw here instead of killing the process

File: src/cpprealm/app.hpp [hunk 7]
```diff
@@ -280,10 +388,11 @@ class App {
 
         m_app = app::App::get_shared_app(app::App::Config{
             .app_id=app_id,
-            .transport = std::make_shared<DefaultTransport>(),
+            .transport = std::make_shared<internal::DefaultTransport>(),
+            .base_url = base_url.empty() ? util::Optional<std::string>() : base_url,
```
Comment (leemaguire) [inline src/cpprealm/app.hpp:392]: why not just make `baas_url` an optional param to begin with?

File: tests/admin_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,6 @@
+//
+// Created by Jason Flax on 9/20/22.
```
Comment (leemaguire) [inline tests/admin_utils.cpp:2]: needs licence header

File: tests/admin_utils.hpp [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+                                     });
+            parsed_response = static_cast<bson::BsonDocument>(bson::parse(result.body));
+            auto roles = static_cast<bson::BsonArray>(parsed_response["roles"]);
+            auto group_id = static_cast<std::string>(static_cast<bson::BsonDocument>(roles[0])["group_id"]);
+
+            Session session = {access_token, group_id};
+            session.apps = Endpoint{access_token,
+                            group_id,
+                            "http://localhost:9090/api/admin/v3.0/groups/" + group_id + "/apps"};
+//            shared = session;
```
Comment (leemaguire) [inline tests/admin_utils.hpp:323]: unused code

File: tests/admin_utils.hpp [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+                                    }}}
+                            }}
+                    }}
+            };
+            app["services"][static_cast<std::string>(service_id)]["config"].patch(std::move(service_config));
+            auto config = app["sync"]["config"];
+            config.put(bson::BsonDocument {{"development_mode_enabled", true}});
+            return static_cast<std::string>(info["client_app_id"]);
+//
+//            app.functions.post(on: group, [
```
Comment (leemaguire) [inline tests/admin_utils.hpp:435]: remove unused code

File: tests/dependencies.list [hunk 11]
```diff
@@ -0,0 +1,3 @@
+VERSION=10.28.7
+REALM_CORE_VERSION=12.5.1
```
Comment (leemaguire) [inline tests/dependencies.list:2]: Prob should bump to core v12.7.0

File: CMakeLists.txt (hunk 1 above)
Comment (jsflax) [reply CMakeLists.txt:7]: Because it should only be on for testing.

File: src/cpprealm/app.hpp (hunk 7 above)
Comment (jsflax) [reply src/cpprealm/app.hpp:390]: Good call.

**Review by leemaguire — APPROVED** [review]

**Pull request merged**